### PR TITLE
HIVE-27157: AssertionError when inferring return type for unix_timestamp function

### PR DESCRIPTION
**Pull request by zabetak**
### What changes were proposed in this pull request?
Change the implementation of unix_timestamp operators to avoid the AssertionError and infer the return type correctly; always BIGINT.

Break the inheritance relation with SqlAbstractTimeFunction and change the SqlFunctionCategory from TIMEDATE to NUMERIC; unix_timestamp is not a time function since the result is never among DATE, TIME, or TIMESTAMP.

Change the operand type checker to a more truthful implementation; the type checker is not really used at the moment but it is better to have something realistic there instead of null or something completely wrong.

Change the function syntax from FUNCTION_ID to FUNCTION and update some .q.out files. Not a must do but again there is no reason to omit parentheses from a regular function.

### Why are the changes needed?
Calls to `inferReturnType` method for `unix_timestamp` operators always lead to `AssertionError`.

Contrary to operand type checking and operand type inference that are not really relevant for Hive (the latter is not using the `SqlValidator` logic), the return type inference is important since it may kick in some calls to `RelBuilder/RexBuilder` APIs. Such calls exist in older versions of Hive and are widely used in Calcite's built-in rules.

### Does this PR introduce _any_ user-facing change?
In this version no, but in older versions of Hive it can fix queries failing with `AssertionError`.

### How was this patch tested?
```
mvn test -Dtest=TestSqlOperatorInferReturnType
```

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/reloperators/HiveToUnixTimestampSqlOperator.java [hunk 1]
```diff
@@ -18,15 +18,24 @@
...
+public class HiveToUnixTimestampSqlOperator {
+  public static final SqlFunction INSTANCE =
+      new SqlFunction("UNIX_TIMESTAMP", SqlKind.OTHER_FUNCTION, ReturnTypes.BIGINT, null,
+          OperandTypes.or(OperandTypes.NILADIC,
+              OperandTypes.or(OperandTypes.STRING, OperandTypes.TIMESTAMP, OperandTypes.DATE),
+              OperandTypes.STRING_STRING), SqlFunctionCategory.NUMERIC) {
+        @Override
+        public boolean isDynamicFunction() {
+          return true;
+        }
```
Comment (kasakrisz) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/reloperators/HiveToUnixTimestampSqlOperator.java:39]: Is this the same constant definition in `HiveUnixTimestampSqlOperator`? Is it worth referencing it from both classes?

Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/reloperators/HiveToUnixTimestampSqlOperator.java:39]: It was the same before and it remains the same now. Ultimately, I would like to get rid one of them completely but it was not trivial to do now (correlation with code in Druid storage handler) thus in order to keep the PR minimal I decided to defer this to HIVE-27162.

Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/reloperators/HiveToUnixTimestampSqlOperator.java:39]: Forgot to mention that there are some places in the code that rely on the fact that there are two different instances.

**Review by kasakrisz — APPROVED** [review]
LGTM

**Pull request closed without merging**